File: crates/algal/src/application_restoration.rs
//! Forward-only restoration of retained pure strategies. Stored records do
//! not authorize restoration; PolicyHost requires an explicit policy option.
use crate::application::{Service, parse_revision, parse_state};
use crate::application_memory::{
    app_id, app_object, app_object_opt, app_ref, app_refs, app_tag, get_record, opt_ref, put_record,
};
use crate::canonical::digest;
use crate::graph::{Transports, compile, interface_signature};
use crate::store::Store;
use crate::{Error, Result};
use serde_json::{Value, json};
use std::collections::BTreeSet;

pub fn parse_policy(input: &Value) -> Result<Value> {
    let v = app_object(input, &["contract", "application", "mode"])?;
    app_tag(&v["contract"], "algal.application-restoration-policy.v1")?;
    app_id(&v["application"])?;
    app_tag(&v["mode"], "retained-pure-strategy-manifests")?;
    Ok(input.clone())
}

pub fn parse_restoration(input: &Value) -> Result<Value> {
    let v = app_object(
        input,
        &[
            "contract",
            "application",
            "parentState",
            "targetState",
            "candidateRevision",
            "policy",
        ],
    )?;
    app_tag(&v["contract"], "algal.application-restoration.v1")?;
    app_id(&v["application"])?;
    for field in ["parentState", "targetState", "candidateRevision", "policy"] {
        app_ref(&v[field])?;
    }
    Ok(input.clone())
}

fn restored_revision(
    store: &Store,
    application: &str,
    parent_state: &str,
    target_state: &str,
) -> Result<Value> {
    let parent = parse_state(&get_record(store, parent_state)?)?;
    if parent.application != application {
        return Err(Error::invalid("Restoration belongs to another application"));
    }
    let mut cursor = parent.previous.clone();
    let mut seen = BTreeSet::from([parent_state.to_owned()]);
    while let Some(reference) = cursor.as_ref() {
        if seen.len() >= 4096 || !seen.insert(reference.clone()) {
            return Err(Error::invalid("Restoration ancestry bound/cycle"));
        }
        let state = parse_state(&get_record(store, reference)?)?;
        if state.application != application {
            return Err(Error::invalid(
                "Restoration ancestry belongs to another application",
            ));
        }
        if reference == target_state {
            break;
        }
        cursor = state.previous;
    }
    if cursor.is_none() {
        return Err(Error::invalid(
            "Restoration target is not a retained ancestor",
        ));
    }
    let target = parse_state(&get_record(store, target_state)?)?;
    let current = parse_revision(&get_record(store, &parent.revision)?)?;
    let historical = parse_revision(&get_record(store, &target.revision)?)?;
    if current.application != application
        || historical.application != application
        || current
            .entrypoints
            .iter()
            .map(|e| &e.name)
            .collect::<Vec<_>>()
            != historical
                .entrypoints
                .iter()
                .map(|e| &e.name)
                .collect::<Vec<_>>()
    {
        return Err(Error::invalid("Restoration entrypoints differ"));
    }
    let mut candidate = current.value.clone();
    candidate["parent"] = json!(parent.revision);
    let mut changed = 0;
    for (index, entry) in current.entrypoints.iter().enumerate() {
        let old = &historical.entrypoints[index];
        if old.manifest != entry.manifest {
            changed += 1;
            let incumbent = store.manifest(&entry.manifest)?;
            let target_manifest = store.manifest(&old.manifest)?;
            for manifest in [&incumbent, &target_manifest] {
                if manifest
                    .cells
                    .iter()
                    .any(|c| !matches!(c["kind"].as_str(), Some("input" | "const" | "fn" | "expr")))
                {
                    return Err(Error::invalid(
                        "Restoration can only change pure strategy manifests",
                    ));
                }
            }
            let old_b = &target_manifest.budgets;
            let now = &incumbent.budgets;
            if old_b.max_steps > now.max_steps
                || old_b.max_agent_calls > now.max_agent_calls
                || old_b.max_work > now.max_work
                || old_b.max_context_bytes > now.max_context_bytes
                || old_b.max_output_bytes > now.max_output_bytes
                || old_b.max_depth > now.max_depth
            {
                return Err(Error::invalid("Restoration widens a manifest budget"));
            }
            let mut overlay = store.overlay();
            let left = interface_signature(&compile(
                incumbent,
                &mut overlay,
                &Default::default(),
                &Transports::new(),
                0,
            )?)?;
            let right = interface_signature(&compile(
                target_manifest,
                &mut overlay,
                &Default::default(),
                &Transports::new(),
                0,
            )?)?;
            if left.inputs != right.inputs || left.outputs != right.outputs {
                return Err(Error::invalid("Restoration changes the compiled interface"));
            }
        }
        candidate["entrypoints"][index]["manifest"] = json!(old.manifest);
    }
    if changed == 0 {
        return Err(Error::invalid(
            "Restoration must change a strategy manifest",
        ));
    }
    parse_revision(&candidate)?;
    Ok(candidate)
}

pub fn verify_restoration(
    store: &Store,
    application: &str,
    parent_state: &str,
    candidate_revision: &str,
    evidence: &[String],
) -> Result<Value> {
    let mut records = Vec::new();
    for reference in evidence {
        let value = get_record(store, reference)?;
        if value["contract"] == "algal.application-restoration.v1" {
            records.push(parse_restoration(&value)?);
        }
    }
    if records.len() != 1 {
        return Err(Error::invalid(
            "Restoration requires exactly one restoration record",
        ));
    }
    let record = records.remove(0);
    if record["application"] != application
        || record["parentState"] != parent_state
        || record["candidateRevision"] != candidate_revision
    {
        return Err(Error::invalid(
            "Restoration evidence does not bind this transition",
        ));
    }
    let policy = parse_policy(&get_record(store, app_ref(&record["policy"])?)?)?;
    if policy["application"] != application {
        return Err(Error::invalid(
            "Restoration policy belongs to another application",
        ));
    }
    let expected = restored_revision(
        store,
        application,
        parent_state,
        app_ref(&record["targetState"])?,
    )?;
    let candidate = parse_revision(&get_record(store, candidate_revision)?)?;
    if digest(&expected)? != digest(&candidate.value)? {
        return Err(Error::invalid(
            "Restoration must preserve current revision metadata and authority",
        ));
    }
    Ok(record)
}

pub async fn restore_revision(lifecycle: &mut Service<'_>, input: &Value) -> Result<Value> {
    let v = app_object_opt(
        input,
        &[
            "application",
            "operation",
            "expectedHead",
            "targetState",
            "policy",
        ],
        &["evidence", "causedBy"],
    )?;
    let application = app_id(&v["application"])?;
    let operation = app_ref(&v["operation"])?;
    let expected_head = app_ref(&v["expectedHead"])?;
    let target_state = app_ref(&v["targetState"])?;
    let policy = app_ref(&v["policy"])?;
    let mut evidence = match v.get("evidence") {
        Some(v) => app_refs(v, 15)?,
        None => Vec::new(),
    };
    let caused_by = opt_ref(v.get("causedBy").unwrap_or(&Value::Null))?;
    let candidate = restored_revision(&lifecycle.store, application, expected_head, target_state)?;
    let revision = put_record(&mut lifecycle.store, &candidate)?;
    let restoration = put_record(
        &mut lifecycle.store,
        &json!({"contract":"algal.application-restoration.v1", "application":application, "parentState":expected_head,"targetState":target_state,"candidateRevision":revision,"policy":policy}),
    )?;
    let current = parse_state(&get_record(&lifecycle.store, expected_head)?)?;
    evidence.push(restoration.clone());
    evidence.sort();
    evidence.dedup();
    let snapshot = lifecycle.commit(&json!({"application":application,"operation":operation,"kind":"restore","expectedHead":expected_head,"revision":revision,"memory":current.memory,"intents":[],"evidence":evidence,"causedBy":caused_by})).await?;
    Ok(json!({"snapshot":snapshot.digest,"revision":revision,"restoration":restoration}))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::contract::Manifest;

    fn strategy(value: &str) -> Manifest {
        Manifest::parse(&json!({"contract":"algal.organism.v1","key":"organism:strategy","name":"Strategy",
            "interface":{"inputs":{},"outputs":{"strategy":{"cell":"out","port":"value"}}},
            "cells":[{"id":"out","kind":"const","outputs":{"value":{"type":"json","value":value}}}],"edges":[]})).unwrap()
    }

    fn fixture(store: &mut Store, incumbent: &Manifest) -> (String, String, String) {
        let reference = put_record(store, &json!("retained reference")).unwrap();
        let prior_manifest = store.put("manifests", &strategy("prior").value).unwrap();
        let current_manifest = store.put("manifests", &incumbent.value).unwrap();
        let revision = json!({"contract":"algal.application-revision.v1","application":"workspace","parent":null,
            "schema":reference,"queries":reference,"views":reference,"runtimeProfile":reference,"evaluationPolicy":reference,
            "capabilityRequirements":[],"entrypoints":[{"name":"strategy","manifest":prior_manifest,"applicability":reference,"queries":[reference],"capabilities":[],"maxGenerations":1}]});
        let prior_revision = put_record(store, &revision).unwrap();
        let state = json!({"contract":"algal.application-state.v1","application":"workspace","sequence":0,"epoch":0,"revision":prior_revision,"memory":reference,"previous":null,"transition":reference});
        let prior_state = put_record(store, &state).unwrap();
        let mut current = revision;
        current["parent"] = json!(prior_revision);
        current["entrypoints"][0]["manifest"] = json!(current_manifest);
        current["views"] = json!(put_record(store, &json!("current view")).unwrap());
        let current_revision = put_record(store, &current).unwrap();
        let mut next = state;
        next["previous"] = json!(prior_state);
        next["revision"] = json!(current_revision);
        next["sequence"] = json!(1);
        next["epoch"] = json!(1);
        let parent_state = put_record(store, &next).unwrap();
        (prior_state, parent_state, current_revision)
    }

    #[test]
    fn restoration_preserves_current_metadata_and_rejects_non_ancestry_or_tampering() {
        let dir = tempfile::tempdir().unwrap();
        let mut store = Store::open(dir.path(), true).unwrap();
        let (prior, parent, current) = fixture(&mut store, &strategy("current"));
        let restored = restored_revision(&store, "workspace", &parent, &prior).unwrap();
        assert_eq!(restored["parent"], current);
        assert_eq!(
            restored["views"],
            get_record(&store, &current).unwrap()["views"]
        );
        assert!(restored_revision(&store, "workspace", &parent, &parent).is_err());
        let policy = put_record(&mut store, &json!({"contract":"algal.application-restoration-policy.v1","application":"workspace","mode":"retained-pure-strategy-manifests"})).unwrap();
        let candidate = put_record(&mut store, &restored).unwrap();
        let record = json!({"contract":"algal.application-restoration.v1","application":"workspace","parentState":parent,"targetState":prior,"candidateRevision":candidate,"policy":policy});
        let evidence = put_record(&mut store, &record).unwrap();
        verify_restoration(&store, "workspace", &parent, &candidate, &[evidence]).unwrap();
        let mut altered = restored;
        altered["goals"] = json!([]);
        let candidate = put_record(&mut store, &altered).unwrap();
        let mut record = record;
        record["candidateRevision"] = json!(candidate);
        let evidence = put_record(&mut store, &record).unwrap();
        assert!(verify_restoration(&store, "workspace", &parent, &candidate, &[evidence]).is_err());
    }

    #[test]
    fn restoration_never_widens_budget_or_changes_effectful_harness() {
        for kind in ["budget", "agent", "interface"] {
            let dir = tempfile::tempdir().unwrap();
            let mut store = Store::open(dir.path(), true).unwrap();
            let mut candidate = strategy("current").value;
            match kind {
                "budget" => candidate["budgets"]["maxWork"] = json!(1),
                "agent" => {
                    candidate["cells"] = json!([{"id":"out","kind":"agent","prompt":"Never executed","output":{"kind":"text"}}])
                }
                _ => {
                    candidate["interface"]["outputs"] =
                        json!({"changed":{"cell":"out","port":"value"}})
                }
            }
            let (prior, parent, _) = fixture(&mut store, &Manifest::parse(&candidate).unwrap());
            assert!(
                restored_revision(&store, "workspace", &parent, &prior).is_err(),
                "{kind}"
            );
        }
        assert!(parse_policy(&json!({"contract":"algal.application-restoration-policy.v1","application":"workspace","mode":"retained-pure-strategy-manifests","extra":true})).is_err());
    }
}

File: src/application-restoration.ts
/** Forward-only restoration of retained pure strategies. Authority is an
 * explicit host option; a stored policy or historical state grants none. */
import {
  APPLICATION_LIMITS, applicationId, applicationJson, applicationObject,
  applicationRef, applicationRefs, applicationTag, getApplicationRecord,
  nullableApplicationRef, parseApplicationRevision, parseApplicationState,
  putApplicationRecord, type ApplicationRevision,
} from "./application-contract";
import type { ApplicationCore, ApplicationSnapshot } from "./application-core";
import { digestCanonical, type Digest } from "./digest";
import { compileOrganism, interfaceSignature } from "./graph";
import { builtinRegistry } from "./registry";
import { replayStore } from "./store-memory";
import type { Store } from "./store-contract";

export type ApplicationRestorationPolicy = {
  contract: "algal.application-restoration-policy.v1";
  application: string;
  mode: "retained-pure-strategy-manifests";
};
export type ApplicationRestoration = {
  contract: "algal.application-restoration.v1";
  application: string;
  parentState: Digest; targetState: Digest; candidateRevision: Digest; policy: Digest;
};
export type RestoreApplicationRevisionInput = {
  application: string; operation: Digest; expectedHead: Digest; targetState: Digest; policy: Digest;
  evidence?: Digest[]; causedBy?: Digest | null;
};
export type RestoredApplicationRevision = { snapshot: ApplicationSnapshot; revision: Digest; restoration: Digest };
const hash = (value: unknown) => digestCanonical(applicationJson(value));

export function parseApplicationRestorationPolicy(input: unknown): ApplicationRestorationPolicy {
  const v = applicationObject(input, ["contract", "application", "mode"]);
  applicationTag(v.contract, "algal.application-restoration-policy.v1");
  applicationTag(v.mode, "retained-pure-strategy-manifests");
  return { contract: "algal.application-restoration-policy.v1", application: applicationId(v.application), mode: "retained-pure-strategy-manifests" };
}
export function parseApplicationRestoration(input: unknown): ApplicationRestoration {
  const v = applicationObject(input, ["contract", "application", "parentState", "targetState", "candidateRevision", "policy"]);
  applicationTag(v.contract, "algal.application-restoration.v1");
  return { contract: "algal.application-restoration.v1", application: applicationId(v.application), parentState: applicationRef(v.parentState), targetState: applicationRef(v.targetState), candidateRevision: applicationRef(v.candidateRevision), policy: applicationRef(v.policy) };
}

async function restoredRevision(store: Store, application: string, parentState: Digest, targetState: Digest): Promise<ApplicationRevision> {
  const parent = await getApplicationRecord(store, parentState, parseApplicationState);
  if (parent.application !== application) throw new Error("Restoration belongs to another application");
  let cursor = parent.previous;
  const seen = new Set<Digest>([parentState]);
  while (cursor !== null) {
    if (seen.size >= APPLICATION_LIMITS.states || seen.has(cursor)) throw new Error("Restoration ancestry bound/cycle");
    seen.add(cursor);
    const state = await getApplicationRecord(store, cursor, parseApplicationState);
    if (state.application !== application) throw new Error("Restoration ancestry belongs to another application");
    if (cursor === targetState) break;
    cursor = state.previous;
  }
  if (cursor === null) throw new Error("Restoration target is not a retained ancestor");
  const target = await getApplicationRecord(store, targetState, parseApplicationState);
  const current = await getApplicationRecord(store, parent.revision, parseApplicationRevision);
  const historical = await getApplicationRecord(store, target.revision, parseApplicationRevision);
  if (current.application !== application || historical.application !== application || hash(current.entrypoints.map(entry => entry.name)) !== hash(historical.entrypoints.map(entry => entry.name))) throw new Error("Restoration entrypoints differ");
  let changed = 0;
  const entrypoints = [];
  for (const entry of current.entrypoints) {
    const old = historical.entrypoints.find(row => row.name === entry.name)!;
    if (old.manifest !== entry.manifest) {
      changed++;
      const incumbent = await store.getManifest(entry.manifest), candidate = await store.getManifest(old.manifest);
      if (!incumbent || !candidate) throw new Error("Restoration manifest is missing");
      for (const manifest of [incumbent, candidate]) if (manifest.cells.some(cell => !["input", "const", "fn", "expr"].includes(cell.kind))) throw new Error("Restoration can only change pure strategy manifests");
      for (const key of ["maxSteps", "maxAgentCalls", "maxWork", "maxContextBytes", "maxOutputBytes", "maxDepth"] as const) if (candidate.budgets[key] > incumbent.budgets[key]) throw new Error("Restoration widens a manifest budget");
      const registry = builtinRegistry(), readOnly = replayStore(store);
      const left = await compileOrganism(incumbent, registry, readOnly), right = await compileOrganism(candidate, registry, readOnly);
      if (hash(interfaceSignature(left)) !== hash(interfaceSignature(right))) throw new Error("Restoration changes the compiled interface");
    }
    entrypoints.push({ ...entry, manifest: old.manifest });
  }
  if (!changed) throw new Error("Restoration must change a strategy manifest");
  return parseApplicationRevision({ ...current, parent: parent.revision, entrypoints });
}

/** Structural verification is independent of host authority and runs on both
 * commit and retained-history inspection. Only the host can authorize policy. */
export async function verifyApplicationRestoration(store: Store, input: {
  application: string; parentState: Digest; candidateRevision: Digest; evidence: Digest[];
}): Promise<ApplicationRestoration> {
  const matches: ApplicationRestoration[] = [];
  for (const reference of input.evidence) {
    const value = await getApplicationRecord(store, reference, applicationJson);
    if (value && typeof value === "object" && !Array.isArray(value) && value.contract === "algal.application-restoration.v1") matches.push(parseApplicationRestoration(value));
  }
  if (matches.length !== 1) throw new Error("Restoration requires exactly one restoration record");
  const record = matches[0]!;
  if (record.application !== input.application || record.parentState !== input.parentState || record.candidateRevision !== input.candidateRevision) throw new Error("Restoration evidence does not bind this transition");
  const policy = await getApplicationRecord(store, record.policy, parseApplicationRestorationPolicy);
  if (policy.application !== input.application) throw new Error("Restoration policy belongs to another application");
  const expected = await restoredRevision(store, record.application, record.parentState, record.targetState);
  const candidate = await getApplicationRecord(store, input.candidateRevision, parseApplicationRevision);
  if (hash(candidate) !== hash(expected)) throw new Error("Restoration must preserve current revision metadata and authority");
  return record;
}

export async function restoreApplicationRevision(lifecycle: ApplicationCore, input: RestoreApplicationRevisionInput): Promise<RestoredApplicationRevision> {
  const optional = ["evidence", "causedBy"].filter(key => Object.hasOwn(input, key));
  const v = applicationObject(input, ["application", "operation", "expectedHead", "targetState", "policy", ...optional]);
  const application = applicationId(v.application), operation = applicationRef(v.operation), expectedHead = applicationRef(v.expectedHead);
  const targetState = applicationRef(v.targetState), policy = applicationRef(v.policy), evidence = applicationRefs(v.evidence ?? [], 15), causedBy = nullableApplicationRef(v.causedBy ?? null);
  const revision = await putApplicationRecord(lifecycle.store, await restoredRevision(lifecycle.store, application, expectedHead, targetState));
  const restoration = await putApplicationRecord(lifecycle.store, { contract: "algal.application-restoration.v1", application, parentState: expectedHead, targetState, candidateRevision: revision, policy });
  const current = await getApplicationRecord(lifecycle.store, expectedHead, parseApplicationState);
  const snapshot = await lifecycle.commit({ application, operation, kind: "restore", expectedHead, revision, memory: current.memory, intents: [], evidence: [...new Set([...evidence, restoration])].sort(), causedBy });
  return { snapshot, revision, restoration };
}
